virtdbg/vmx/vmm.c

```diff
@@ -0,0 +1,293 @@
+#include <vmx/vmm.h>
+#include <mm/pmm.h>
+#include <arch/intrin.h>
+#include <stdint.h>
+#include <vmx/ept.h>
+#include <stddef.h>
+#include <util/except.h>
+#include <virtdbg.h>
+
+extern void vm_resume(guest_state_t *t);
+extern ept_entry_t* m_root_pa;
+
+//enable vmx operation
+err_t vmxon() {
+    err_t err = NO_ERROR;
+    uint32_t* vmxon_region = pallocz_aligned(0x1000, 0x1000);
+
+    size_t control = __rdmsr(0x3a);
+    if((control & (0x1 | 0x4)) != (0x1 | 0x04)) {
+        //Enabled outside of SMX and lock bit.
+        __wrmsr(0x3a, control | 0x1 | 0x4);
+    }
+
+    uint64_t cr0 = __readcr0().raw;
+    cr0 &= __rdmsr(0x487);
+    cr0 |= __rdmsr(0x486);
+    __writecr0((ia32_cr0_t) { .raw = cr0 });
+
+    uint64_t cr4;
+    asm volatile ("mov %%cr4, %0" : "=r" (cr4));
```

Comment:
You have functions for cr4, and if you want the raw value they are union for a reason